10-JavaScript/arrow-functions.md

```diff
@@ -0,0 +1,261 @@
+
+### Converting a function into an arrow function
+
+Traditional functions look like this:
+
+```javascript
+const foo = function() {
+  return "bar";
+}
+
+console.log(foo());  
+> bar
+```
+
+Using an arrow function we could reduce it as:
+
+```javascript
+// if there are no parameters you need to include empty parentheses to indicate it's a function
+const foo = () => {
+  return "bar";
+}
+
+console.log(foo());  
+> bar
+```
+
+So what have we done?  We've gotten rid of the `function` keyword and placed a "arrow" to the right of the function parameters.  This lets us create a function with less typing, but we can do this with even less code.
+
+**Note**:  The arrow must be on the same line as the parentheses.  If they are on separate lines, it will generate a Syntax error.
+
+Because our function has only a **one** line return, we can remove the return keyword and make it smaller.
+
+```javascript
+const foo = () => {  "bar"; }
+
+console.log(foo());  
+>  bar
+```
+
+Since `"bar"` is the only line in the function, it's considered an _implicit return_.  In other words, JavaScript assumes the function will return that value.  That's a lot less code, but we can reduce it further.  Since this is a 1-line function we can remove the curly braces and make them implicit as well.
+
+```javascript
+const foo = () => "bar";
+
+console.log(foo());  
+>  bar
+```
+
+Wow that's a lot less code.
+
+### Exercise
+
+Here is an example for defining a function with no parameters which returns a string "hello world".  Convert this into an arrow function and minimize the code typed.
+
+```javascript
+const hello = function() {
+  return "hello world"
+};
+```
+
+### Arrow Functions with parameters
+
+```javascript
+const squarer = (number) => {
+  return number * number
+}
+```
+This arrow function takes a single number and returns that number squared.  Like normal functions arrow functions have their parameters listed in the parentheses.  
+
+So this is a pretty compact syntax, but... we're not done.  Because this function takes a **single** argument, we can dispense with the parentheses around `number`.
+
+```javascript
+const squarer = number => {return number * number;};
+squared([3, 4, 5]);
+> [ 9, 16, 25 ]
+```
+
+**Note**: This works **only** for functions with a single argument.  Zero parameters or more than one will require you to include the parentheses.  
+
+For example with two parameters
+
+```javascript
+const add = (x, y) => x + y;
+```
+
+Just like earlier since `squarer` has a 1-line function body we can remove the curly braces.
+
+```javascript
+const squarer = number => return number * number; ;
+squared([3, 4, 5]);
+> [ 9, 16, 25 ]
+```
```

Comment:
I agree it's terrible, but since it was in a bunch of ES6 references, I put it in.  

I'm happy to either take it out or leave a note.  I'm just going through a React tutorial now that uses this syntax all the time, it still kills me.

@kariabancroft any opinions?

Reply:
I agree it looks terrible and also agree that despite that they're going to see it everywhere.

Reply:
OK so leave a note.

Reply:
Note added